server/src/main/java/org/elasticsearch/indices/recovery/MultiFileTransfer.java

```diff
@@ -0,0 +1,211 @@
+ * checkpoint tracker for this purpose. We generate a new sequence number and assign it to each file chunk request before sending; then mark
+ * that sequence number as processed when we receive a response for the corresponding file chunk request. With the local checkpoint tracker,
+ * we know the last acknowledged-flushed file-chunk is a file chunk whose {@code requestSeqId} equals to the local checkpoint because the
+ * recover target can flush all file chunks up to the local checkpoint.
+ * <p>
+ * When the number of un-replied file chunk requests reaches the limit (i.e. the gap between the max_seq_no and the local checkpoint is
+ * greater than {@code maxConcurrentFileChunks}), the sending/requesting thread will abort its execution. That process will be resumed by
+ * one of the networking threads which receive/handle the responses of the current pending file chunk requests. This process will continue
+ * until all chunk requests are sent/responded.
+ */
+abstract class MultiFileTransfer<Request extends MultiFileTransfer.ChunkRequest> implements Closeable {
+    private Status status = Status.PROCESSING;
+    private final Logger logger;
+    private final ActionListener<Void> listener;
+    private final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+    private final AsyncIOProcessor<FileChunkResponseItem> processor;
+    private final int maxConcurrentFileChunks;
+    private StoreFileMetaData currentFile = null;
+    private final Iterator<StoreFileMetaData> remainingFiles;
+    private Tuple<StoreFileMetaData, Request> readAheadRequest = null;
+
+    protected MultiFileTransfer(Logger logger, ThreadContext threadContext, ActionListener<Void> listener,
+                                int maxConcurrentFileChunks, List<StoreFileMetaData> files) {
+        this.logger = logger;
+        this.maxConcurrentFileChunks = maxConcurrentFileChunks;
+        this.listener = listener;
+        this.processor = new AsyncIOProcessor<>(logger, maxConcurrentFileChunks, threadContext) {
+            @Override
+            protected void write(List<Tuple<FileChunkResponseItem, Consumer<Exception>>> items) {
+                handleItems(items);
+            }
+        };
+        this.remainingFiles = files.iterator();
+    }
+
+    public final void start() {
+        addItem(UNASSIGNED_SEQ_NO, null, null); // put a dummy item to start the processor
+    }
+
+    private void addItem(long requestSeqId, StoreFileMetaData md, Exception failure) {
+        processor.put(new FileChunkResponseItem(requestSeqId, md, failure), e -> { assert e == null : e; });
+    }
+
+    private void handleItems(List<Tuple<FileChunkResponseItem, Consumer<Exception>>> items) {
+        if (status != Status.PROCESSING) {
+            assert status != Status.SUCCESS : "must not receive any response after the transfer was completed";
```

Comment:
perhaps `assert status == FAILED` to make it clearer what state we're supposed to be in

Reply:
addressed in https://github.com/elastic/elasticsearch/commit/6d768948069f004c67d98b4e050e4811018605dc